server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -1149,6 +1149,27 @@ public <T> List<T> readList(final Writeable.Reader<T> reader) throws IOException
         return readCollection(reader, ArrayList::new, Collections.emptyList());
     }
 
+    /**
+     * Reads an list of objects. The list is expected to have been written using {@link StreamOutput#writeList(List)}.
+     * The returned list is immutable.
+     *
+     * @return the list of objects
+     * @throws IOException if an I/O exception occurs reading the list
+     */
+    public <T> List<T> readImmutableList(final Writeable.Reader<T> reader) throws IOException {
+        int count = readArraySize();
+        if (count == 0) {
```

Comment:
I think it makes sense to special case `count == 1` and `== 2` as well here. These are quite common and the JDK does the same in `java.util.List#of(E...)` so I'd assume that's the right level of granularity?